{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TTdFB5Jyj0lv"
      },
      "source": [
        "# Final Project of Introduction to Bioinformatics\n",
        "\n",
        "## Comparative Mitochondrial Genomics and Phylogenetic Analysis\n",
        "\n",
        "### Please Note That the Project Is Located at https://colab.research.google.com/drive/1eENAo7y5OAmH4wtYYULIDm00sx3C4m9y?usp=sharing"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6IS6UPYvj0lx"
      },
      "source": [
        "This task focuses on the comparative analysis of mitochondrial genomes from different species, primarily birds, mammals, and insects. The aim is to understand the evolutionary relationships between these species by analyzing and comparing their mitochondrial DNA, which is about 16,000 base pairs in length. You will use advanced computational methods to construct phylogenetic trees and delve into the ecological and anthropological insights that can be gleaned from this data. This project is designed to provide a comprehensive understanding of mitochondrial genomics, its importance in evolutionary biology, and its applications in broader scientific contexts.\n",
        "\n",
        "You will learn:\n",
        "\n",
        "- Techniques for aligning and comparing mitochondrial DNA sequences.\n",
        "- How to construct and interpret phylogenetic trees using advanced computational methods.\n",
        "- The application of mitochondrial genomics in understanding ecological interactions and human evolutionary history.\n",
        "\n",
        "#### Task Roadmap\n",
        "\n",
        "1. **Mitochondrial Genome Comparison**:\n",
        "   - Align mitochondrial DNA sequences from the provided dataset.\n",
        "   - Analyze these sequences to identify similarities and differences across species.\n",
        "\n",
        "2. **Phylogenetic Analysis Using Advanced Methods**:\n",
        "   - Apply Maximum Likelihood (ML) and Bayesian Inference methods, utilizing tools like `ETE Toolkit`, `DendroPy`, `BEAST`, or `PyRate`.\n",
        "   - Compare the trees generated by these methods to understand how different approaches can lead to different interpretations of the data.\n",
        "\n",
        "3. **Cross-Disciplinary Applications (Bonus)**:\n",
        "   - **Ecology**: Examine how mitochondrial DNA analysis can reveal information about species adaptation, migration, and conservation. This involves understanding how genetic variation within and between species can inform ecological strategies and conservation efforts.\n",
        "   - **Anthropology**: Investigate the use of mitochondrial DNA in tracing human evolution and migration patterns. This includes studying the mitochondrial DNA of mammals in your dataset to draw parallels with human evolutionary studies.\n",
        "\n",
        "### Data Sources\n",
        "\n",
        "The mitochondrial DNA data for birds, mammals, and insects will be provided to you. This dataset has been curated to facilitate a comprehensive comparative analysis and is essential for the completion of your phylogenetic studies.\n",
        "\n",
        "### Useful Resources and Material\n",
        "\n",
        "- [Mitochondrial DNA - Wikipedia](https://en.wikipedia.org/wiki/Mitochondrial_DNA): A general introduction to the structure, function, origin, and diversity of mitochondrial DNA, as well as its applications in various fields such as medicine, forensics, and anthropology.\n",
        "- [Mitochondrial DNA Analysis: Introduction, Methods, and Applications](https://bioinfo.cd-genomics.com/mitochondrial-dna-analysis-introduction-methods-and-applications.html): An explanation of the basics of mitochondrial DNA sequencing, bioinformatics analysis, heteroplasmy, and advantages of mitochondrial DNA analysis over nuclear DNA analysis.\n",
        "- [Phylogenetic Tree- Definition, Types, Steps, Methods, Uses - Microbe Notes](https://microbenotes.com/phylogenetic-tree/): A coverage of the concepts and methods of phylogenetic tree construction, including the types of phylogenetic trees, the steps involved in phylogenetic analysis, the main methods of phylogenetic inference, and the applications of phylogenetic trees in various disciplines.\n",
        "- [Phylogenetics - Wikipedia](https://en.wikipedia.org/wiki/Phylogenetics): An overview of the field of phylogenetics, which is the study of the evolutionary history and relationships among or within groups of organisms. It also discusses the data sources, models, algorithms, software, and challenges of phylogenetic analysis.\n",
        "- [ETE Toolkit](http://etetoolkit.org/): A Python library for manipulating, analyzing, and visualizing phylogenetic trees. It supports various formats, methods, and tools for phylogenetic analysis, such as alignment, tree inference, tree comparison, tree annotation, and tree visualization.\n",
        "- [DendroPy](https://dendropy.org/): Another Python library for phylogenetic computing. It provides a comprehensive API for working with phylogenetic data structures, such as trees, characters, and networks. It also offers a rich set of functions for simulation, manipulation, analysis, and annotation of phylogenetic data."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MPo_0Wx0j0lz"
      },
      "source": [
        "### Exploration and Reflection\n",
        "\n",
        "As we proceed with our analysis of mitochondrial DNA for phylogenetic tree construction, it is valuable to contemplate a few questions. These inquiries aim to facilitate a more thorough understanding of the roles and characteristics of mitochondrial DNA in the context of evolutionary biology:\n",
        "\n",
        "1. **Maternal Inheritance and Its Implications**: How does the maternal inheritance of mitochondrial DNA simplify our understanding of evolutionary lineage compared to nuclear DNA, which undergoes recombination? What unique insights can this aspect provide in tracing the evolutionary history of species?\n",
        "\n",
        "2. **Mutation Rate and Evolutionary Insights**: Mitochondrial DNA mutates at a faster rate than nuclear DNA. How does this characteristic make mtDNA a more sensitive tool for detecting recent evolutionary events and relationships among closely related species? Can you think of any specific scenarios or studies where this property of mtDNA has been particularly instrumental?\n",
        "\n",
        "Reflect on these questions as you work through the project, and consider how the properties of mitochondrial DNA enhance its value and applicability in evolutionary biology and beyond. Provide your answer either in this notebook, or in your report (if you had one).\n",
        "\n",
        "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
        "    <strong>Don't forget to answer these questions!</strong>\n",
        "</blockquote>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fdxjoSByj0l0"
      },
      "source": [
        "### Step 0: Installing Necessary Packages"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_0WWQ2gyj0l0",
        "outputId": "c409227f-c2f5-4e30-b71f-7b1d8c2f3b62"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "biopython is NOT installed\n",
            "biopython was successfully installed.\n",
            "pandas (1.5.3) is installed\n",
            "numpy (1.23.5) is installed\n"
          ]
        }
      ],
      "source": [
        "import sys\n",
        "import subprocess\n",
        "import pkg_resources\n",
        "\n",
        "def install(package):\n",
        "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", package])\n",
        "\n",
        "REQUIRED_PACKAGES = [\n",
        "    'biopython',\n",
        "    'pandas',\n",
        "    'numpy'\n",
        "]\n",
        "\n",
        "for package in REQUIRED_PACKAGES:\n",
        "    try:\n",
        "        dist = pkg_resources.get_distribution(package)\n",
        "        print('{} ({}) is installed'.format(dist.key, dist.version))\n",
        "    except pkg_resources.DistributionNotFound:\n",
        "        print('{} is NOT installed'.format(package))\n",
        "        install(package)\n",
        "        print('{} was successfully installed.'.format(package))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vA0kgLnRj0l2"
      },
      "outputs": [],
      "source": [
        "# Import necessary libraries.\n",
        "import pandas as pd\n",
        "import numpy as np"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XX-edEISj0l3"
      },
      "source": [
        "### Step 1: Dataset Expansion\n",
        "\n",
        "Our first task is to augment our dataset with additional species. This involves engaging with the NCBI database to retrieve mitochondrial DNA sequences.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Species Selection**: Identify and choose 10 additional species to include in your dataset. Aim for a diverse selection to enrich your phylogenetic analysis.\n",
        "\n",
        "- **Querying NCBI Database**: Use the NCBI database to locate mitochondrial DNA sequences for your chosen species. While you can manually search on the [NCBI website](https://www.ncbi.nlm.nih.gov/), consider automating this process through their API for a more efficient approach.\n",
        "    - **Example Query**: As a starting point, you might use a query like `\"mitochondrion[Filter] AND (your_species_name[Organism])` to find specific mtDNA sequences. Adjust the query parameters according to your species selection.\n",
        "    - **Documentation**: Familiarize yourself with the [NCBI API documentation](https://www.ncbi.nlm.nih.gov/books/NBK25497/) for detailed guidance on constructing queries.\n",
        "\n",
        "- **Using NCBI Website**: You are welcome to use the NCBI website for this task. If you do so, document each step of your process clearly in your task report. This should include the species names, search terms used, and how you determined the relevant sequences to include.\n",
        "\n",
        "- **Bonus Opportunity**: Implementing an automated, methodological approach using the NCBI API and relevant Python packages to add all 10 records in your dataset will earn you a 50% bonus for this section. Your method should be structured and replicable, demonstrating a systematic approach to data collection.\n",
        "\n",
        "Remember, the goal is to methodically expand your dataset with relevant mtDNA sequences, paving the way for insightful phylogenetic analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "unNRMWEcj0l3",
        "outputId": "d33bd65e-fb2a-424b-9293-bf16caad909c"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-effeb4f6-09a9-4548-b7fb-1ebf7af0dd66\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>taxo_id</th>\n",
              "      <th>specie</th>\n",
              "      <th>blast_name</th>\n",
              "      <th>genbank_common_name</th>\n",
              "      <th>accession_number</th>\n",
              "      <th>mtDNA</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8945</td>\n",
              "      <td>Eudynamys scolopaceus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Asian koel</td>\n",
              "      <td>NC_060520</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>7460</td>\n",
              "      <td>Apis mellifera</td>\n",
              "      <td>Bees</td>\n",
              "      <td>honey bee</td>\n",
              "      <td>NC_051932</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>36300</td>\n",
              "      <td>Pelecanus crispus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Dalmatian pelican</td>\n",
              "      <td>OR620163</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>10116</td>\n",
              "      <td>Rattus norvegicus</td>\n",
              "      <td>Rodents</td>\n",
              "      <td>Norway rat</td>\n",
              "      <td>NC_001665</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001665</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>9031</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>NC_053523</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-effeb4f6-09a9-4548-b7fb-1ebf7af0dd66')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-effeb4f6-09a9-4548-b7fb-1ebf7af0dd66 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-effeb4f6-09a9-4548-b7fb-1ebf7af0dd66');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-db4ec791-40b6-4cc1-8312-478f1776a71c\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-db4ec791-40b6-4cc1-8312-478f1776a71c')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-db4ec791-40b6-4cc1-8312-478f1776a71c button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "   taxo_id                 specie blast_name genbank_common_name  \\\n",
              "0     8945  Eudynamys scolopaceus      birds          Asian koel   \n",
              "1     7460         Apis mellifera       Bees           honey bee   \n",
              "2    36300      Pelecanus crispus      birds   Dalmatian pelican   \n",
              "3    10116      Rattus norvegicus    Rodents          Norway rat   \n",
              "4     9031          Gallus gallus      birds       Gallus gallus   \n",
              "\n",
              "  accession_number                                              mtDNA  \n",
              "0        NC_060520  https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...  \n",
              "1        NC_051932  https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...  \n",
              "2         OR620163    https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1  \n",
              "3        NC_001665     https://www.ncbi.nlm.nih.gov/nuccore/NC_001665  \n",
              "4        NC_053523   https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1  "
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "dataset = pd.read_csv('./dataset/species.csv')\n",
        "dataset.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "KOmUsAGWj0l3",
        "outputId": "6bc7dcf3-4f8f-418d-d566-a4f17d82f0b2"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-3b54275a-2baf-4cf6-a38e-8c70cb459baa\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>taxo_id</th>\n",
              "      <th>specie</th>\n",
              "      <th>blast_name</th>\n",
              "      <th>genbank_common_name</th>\n",
              "      <th>accession_number</th>\n",
              "      <th>mtDNA</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>36300</td>\n",
              "      <td>Pelecanus crispus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Dalmatian pelican</td>\n",
              "      <td>OR620163</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>59729</td>\n",
              "      <td>Taeniopygia guttata</td>\n",
              "      <td>birds</td>\n",
              "      <td>zebra finch</td>\n",
              "      <td>NC_007897</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>8845</td>\n",
              "      <td>Anser cygnoides</td>\n",
              "      <td>birds</td>\n",
              "      <td>Swan goose</td>\n",
              "      <td>NC_023832</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>193005</td>\n",
              "      <td>Aquila nipalensis</td>\n",
              "      <td>hawks &amp; eagles</td>\n",
              "      <td>steppe eagle</td>\n",
              "      <td>NC_045042</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>9940</td>\n",
              "      <td>Ovis aries</td>\n",
              "      <td>sheep</td>\n",
              "      <td>even-toed ungulates</td>\n",
              "      <td>NC_001941</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001941</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3b54275a-2baf-4cf6-a38e-8c70cb459baa')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-3b54275a-2baf-4cf6-a38e-8c70cb459baa button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-3b54275a-2baf-4cf6-a38e-8c70cb459baa');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-8ee7aa56-6c3b-42bd-aa24-ec5354dc504f\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-8ee7aa56-6c3b-42bd-aa24-ec5354dc504f')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-8ee7aa56-6c3b-42bd-aa24-ec5354dc504f button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "    taxo_id               specie      blast_name  genbank_common_name  \\\n",
              "2     36300    Pelecanus crispus           birds    Dalmatian pelican   \n",
              "21    59729  Taeniopygia guttata           birds          zebra finch   \n",
              "7      8845      Anser cygnoides           birds           Swan goose   \n",
              "19   193005    Aquila nipalensis  hawks & eagles         steppe eagle   \n",
              "8      9940           Ovis aries           sheep  even-toed ungulates   \n",
              "\n",
              "   accession_number                                              mtDNA  \n",
              "2          OR620163    https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1  \n",
              "21        NC_007897  https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...  \n",
              "7         NC_023832  https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...  \n",
              "19        NC_045042   https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1  \n",
              "8         NC_001941     https://www.ncbi.nlm.nih.gov/nuccore/NC_001941  "
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import pandas as pd\n",
        "import requests\n",
        "\n",
        "# TODO: Add 10 more species to your dataset\n",
        "# for species in additional_species:\n",
        "    # Fetch mtDNA and add to the dataset\n",
        "\n",
        "# TODO: Save the expanded dataset\n",
        "# dataset.to_csv('./dataset/expanded_dataset.csv', index=False)\n",
        "expanded_dataset = pd.read_csv('./dataset/expanded_dataset.csv')\n",
        "expanded_dataset.sample(5)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FZnCFR-tj0l4"
      },
      "source": [
        "#### Checking Data Consistency\n",
        "\n",
        "The code block is designed to check the accuracy of a biological dataset. It examines each entry in a CSV file, focusing on taxonomy IDs, species names, GenBank accession numbers, and mitochondrial DNA links. It uses the NCBI's Entrez system to ensure taxonomy IDs correspond to the correct species and confirms the mitochondrial DNA links are accurate. The script also checks GenBank accession numbers against the provided links. This method is useful for maintaining the accuracy of current data and **might help in adding new entries to the database.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "L8lYlzLcj0l4"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import requests\n",
        "from Bio import Entrez\n",
        "import os\n",
        "from io import BytesIO\n",
        "from requests.adapters import HTTPAdapter\n",
        "from urllib3.util.retry import Retry\n",
        "\n",
        "# Set your email and API key for NCBI\n",
        "entrez_email = os.getenv('ENTREZ_EMAIL')\n",
        "entrez_key = os.getenv('ENTREZ_API_KEY')\n",
        "Entrez.email = entrez_email\n",
        "\n",
        "def fetch_entrez_record(db, id, rettype, retmode):\n",
        "    \"\"\"Fetch record from NCBI Entrez with retries.\"\"\"\n",
        "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={id}&rettype={rettype}&retmode={retmode}\"\n",
        "    session = requests.Session()\n",
        "    retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])\n",
        "    session.mount('https://', HTTPAdapter(max_retries=retries))\n",
        "\n",
        "    try:\n",
        "        response = session.get(url)\n",
        "        response.raise_for_status()\n",
        "        if db == \"nucleotide\":\n",
        "            return response.text\n",
        "        return response.content\n",
        "    except requests.exceptions.HTTPError as err:\n",
        "        print(f\"HTTP error: {err}\")\n",
        "    except requests.exceptions.ConnectionError as err:\n",
        "        print(f\"Connection error: {err}\")\n",
        "    return None\n",
        "\n",
        "def verify_taxonomy_id(taxo_id, species_name):\n",
        "    \"\"\"Verify taxonomy ID against species name.\"\"\"\n",
        "    xml_data = fetch_entrez_record(\"taxonomy\", taxo_id, \"xml\", \"xml\")\n",
        "    if xml_data:\n",
        "        # Convert bytes data to a binary file-like object\n",
        "        xml_data_io = BytesIO(xml_data)\n",
        "        records = Entrez.read(xml_data_io)\n",
        "        return records[0]['ScientificName'].lower() == species_name.lower()\n",
        "    return False\n",
        "\n",
        "def verify_mitochondrial_dna(accession_number):\n",
        "    gb_data = fetch_entrez_record(\"nucleotide\", accession_number, \"gb\", \"text\")\n",
        "    return \"mitochondrion\" in gb_data.lower() if gb_data else False\n",
        "\n",
        "def extract_accession_from_link(link):\n",
        "    return link.split('/')[-1].split('.')[0]\n",
        "\n",
        "def check_dataset_consistency(file_path):\n",
        "    species_df = pd.read_csv(file_path)\n",
        "\n",
        "    for index, row in species_df.iterrows():\n",
        "        taxonomy_id = str(row['taxo_id'])\n",
        "        species_name = row['specie']\n",
        "        accession_number = row['accession_number']\n",
        "        mtDNA_link = row['mtDNA']\n",
        "        extracted_accession = extract_accession_from_link(mtDNA_link)\n",
        "\n",
        "        taxonomy_check = verify_taxonomy_id(taxonomy_id, species_name)\n",
        "        accession_match = (accession_number == extracted_accession)\n",
        "        mitochondrial_check = verify_mitochondrial_dna(accession_number)\n",
        "\n",
        "        print(f\"Row {index}: Taxonomy Check: {taxonomy_check}, Accession Match: {accession_match}, Mitochondrial Check: {mitochondrial_check}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KfU1PpCqj0l5",
        "outputId": "29793134-4e17-411d-c655-48201e5b5015"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Row 0: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 1: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 2: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 3: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 4: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 5: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 6: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 7: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 8: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 9: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 10: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 11: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 12: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 13: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 14: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 15: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 16: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 17: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 18: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 19: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 20: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 21: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 22: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 23: Taxonomy Check: False, Accession Match: True, Mitochondrial Check: True\n",
            "Row 24: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 25: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 26: Taxonomy Check: False, Accession Match: True, Mitochondrial Check: True\n",
            "Row 27: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 28: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 29: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 30: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 31: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 32: Taxonomy Check: False, Accession Match: True, Mitochondrial Check: True\n"
          ]
        }
      ],
      "source": [
        "check_dataset_consistency('./dataset/expanded_dataset.csv')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "d1hR_q3Qj0l5"
      },
      "source": [
        "### Step 3: Sequence Download and Preparation\n",
        "\n",
        "The next step in our project involves downloading the mitochondrial DNA sequences for each species and preparing them for analysis.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Download mtDNA Sequences**: Write a script to download the mtDNA sequences from the links provided in your dataset. The sequences should be in FASTA format, which is the standard for nucleotide sequences.\n",
        "\n",
        "- **Sequence Labeling**: Properly label each sequence within the FASTA file. This header, starting with '>', should include the species name and any other relevant information (e.g., `>Eudynamys_scolopaceus_NC_060520`). This is crucial for identifying the sequences in subsequent analysis.\n",
        "\n",
        "- **Concatenate Sequences**:\n",
        "    - Create a script to concatenate all downloaded sequences into a single `.fasta` or `.fna` file.\n",
        "    - Ensure each sequence in the file is clearly separated by its header line, which is important for differentiating the sequences of various species.\n",
        "\n",
        "#### Tips for Writing the Download and Concatenation Script:\n",
        "- Use Python libraries such as `httpx` or `requests`, or any other tool you prefer for downloading sequences. For processing FASTA files you can use a wide range of tools. One recommended option is `Biopython` library.\n",
        "- Use a loop to go through each link in the dataset, download the sequence, and append it to your concatenated file.\n",
        "- Maintain the format integrity of the FASTA file, ensuring each sequence is correctly associated with its header.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wFmrgvnvj0l6"
      },
      "outputs": [],
      "source": [
        "# TODO: Write a function to download mtDNA sequences\n",
        "# def download_mtDNA(url, label):\n",
        "    # Code to download and label the sequence\n",
        "\n",
        "# TODO: Loop through the dataset and download each mtDNA sequence\n",
        "# for index, row in dataset.iterrows():\n",
        "    # Call the download function for each species"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xl-57qI5IZAc"
      },
      "outputs": [],
      "source": [
        "from Bio import Entrez, SeqIO\n",
        "import pandas as pd\n",
        "\n",
        "def download_mtDNA(accession_number, label):\n",
        "    try:\n",
        "        handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"fasta\", retmode=\"text\")\n",
        "        record = SeqIO.read(handle, \"fasta\")\n",
        "        handle.close()\n",
        "\n",
        "        header = f'>{label}\\n'\n",
        "        return header + str(record.seq)\n",
        "    except Exception as e:\n",
        "        print(f\"Failed to download sequence from {accession_number}: {str(e)}\")\n",
        "        return None\n",
        "\n",
        "def download_concatenate_sequences(dataset):\n",
        "    concatenated_sequences = \"\"\n",
        "\n",
        "    for index, row in dataset.iterrows():\n",
        "        specie_label = row[\"specie\"].replace(\" \", \"_\")\n",
        "        blast_name_label = row[\"blast_name\"].replace(\" \", \"_\")  # Replace spaces with underscores\n",
        "        label = f'{specie_label}_{blast_name_label}_{row[\"accession_number\"]}'\n",
        "        mtDNA_sequence = download_mtDNA(row[\"accession_number\"], label)\n",
        "\n",
        "        if mtDNA_sequence:\n",
        "            concatenated_sequences += mtDNA_sequence + '\\n'\n",
        "\n",
        "    return concatenated_sequences\n",
        "\n",
        "def save_to_fasta(concatenated_sequences, output_file='/content/dataset/concatenated_seq.fasta'):\n",
        "    with open(output_file, 'w') as fasta_file:\n",
        "        fasta_file.write(concatenated_sequences)\n",
        "\n",
        "Entrez.email = 'example@gmail.com'\n",
        "\n",
        "dataset = pd.read_csv('/content/dataset/expanded_dataset.csv')\n",
        "\n",
        "concatenated_sequences = download_concatenate_sequences(dataset)\n",
        "save_to_fasta(concatenated_sequences)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Bvv8hcHsj0l6"
      },
      "source": [
        "### Step 4: Sequence Alignment\n",
        "\n",
        "After downloading the mitochondrial DNA sequences, the next critical step is their alignment. This process allows us to compare the sequences and discern the evolutionary relationships among the species.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Select an Alignment Tool**: Choose one of the following alignment tools based on your project needs. Each tool has its strengths and is widely used in bioinformatics for multiple sequence alignment.\n",
        "\n",
        "1. **MAFFT**:\n",
        "    - **Brief Introduction**: MAFFT (Multiple Alignment using Fast Fourier Transform) is renowned for its speed and efficiency, particularly suitable for large datasets.\n",
        "    - **Resources**:\n",
        "        - [MAFFT Official Documentation](https://mafft.cbrc.jp/alignment/software/)\n",
        "        - [Example Usage on GitHub](https://github.com/MountainMan12/SARS-Cov2-phylo)\n",
        "        - [Relevant Notebook](https://colab.research.google.com/github/pb3lab/ibm3202/blob/master/tutorials/lab03_phylo.ipynb)\n",
        "\n",
        "2. **Clustal Omega**:\n",
        "    - **Brief Introduction**: Clustal Omega offers high-quality alignments and is user-friendly, ideal for those new to sequence alignment.\n",
        "    - **Resources**:\n",
        "        - [A Python wrapper around Clustal Omega](https://github.com/benchling/clustalo-python)\n",
        "        - [Clustal Omega Official Website](http://www.clustal.org/omega/)\n",
        "\n",
        "3. **MUSCLE**:\n",
        "    - **Brief Introduction**: MUSCLE (Multiple Sequence Comparison by Log-Expectation) is known for its balance between speed and accuracy, making it a versatile choice for various datasets.\n",
        "    - **Resources**:\n",
        "        - [MUSCLE Documentation](https://drive5.com/muscle5/manual/)\n",
        "\n",
        "- **Perform Sequence Alignment**: Utilize your chosen tool to align the downloaded mtDNA sequences. This alignment is foundational for the accurate construction of phylogenetic trees.\n",
        "\n",
        "- **Save Aligned Sequences**: After alignment, save the output in an appropriate format for further analysis in the subsequent steps of this project."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BIGnrplTj0l7"
      },
      "outputs": [],
      "source": [
        "# TODO: Install and import the alignment tool\n",
        "# !pip install mafft\n",
        "# from Bio.Align.Applications import MafftCommandline\n",
        "\n",
        "# TODO: Perform the sequence alignment\n",
        "# def perform_alignment(input_file, output_file):\n",
        "    # Code to align sequences using the chosen tool\n",
        "\n",
        "# TODO: Align your downloaded sequences\n",
        "# perform_alignment('path_to_downloaded_sequences.fasta', 'aligned_sequences.fasta')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "no4cVVbLJXVq",
        "outputId": "fa151cce-d357-427c-b5c1-8d576223c6bd"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libargtable2-0\n",
            "The following NEW packages will be installed:\n",
            "  clustalo libargtable2-0\n",
            "0 upgraded, 2 newly installed, 0 to remove and 32 not upgraded.\n",
            "Need to get 273 kB of archives.\n",
            "After this operation, 694 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libargtable2-0 amd64 13-1.1 [14.1 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 clustalo amd64 1.2.4-7 [259 kB]\n",
            "Fetched 273 kB in 0s (911 kB/s)\n",
            "Selecting previously unselected package libargtable2-0.\n",
            "(Reading database ... 121730 files and directories currently installed.)\n",
            "Preparing to unpack .../libargtable2-0_13-1.1_amd64.deb ...\n",
            "Unpacking libargtable2-0 (13-1.1) ...\n",
            "Selecting previously unselected package clustalo.\n",
            "Preparing to unpack .../clustalo_1.2.4-7_amd64.deb ...\n",
            "Unpacking clustalo (1.2.4-7) ...\n",
            "Setting up libargtable2-0 (13-1.1) ...\n",
            "Setting up clustalo (1.2.4-7) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ],
      "source": [
        "!apt-get install -y clustalo"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eZFuP5cU4vGT",
        "outputId": "34216b6e-3fdd-4427-d29b-1e429184da29"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Clustal Omega alignment successful. Aligned sequences saved to /content/dataset/aligned_sequences_clustalo.fasta\n"
          ]
        }
      ],
      "source": [
        "import subprocess\n",
        "\n",
        "# Define the input and output file paths\n",
        "input_file_path = '/content/dataset/concatenated_seq.fasta'\n",
        "output_file_path = '/content/dataset/aligned_sequences_clustalo.fasta'\n",
        "\n",
        "# Run Clustal Omega using subprocess\n",
        "clustalo_cmd = ['clustalo', '--infile', input_file_path, '--outfile', output_file_path, '--outfmt', 'fasta']\n",
        "\n",
        "try:\n",
        "    subprocess.run(clustalo_cmd, check=True)\n",
        "    print(f\"Clustal Omega alignment successful. Aligned sequences saved to {output_file_path}\")\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error during Clustal Omega alignment: {e}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WJS4wd9rj0l7"
      },
      "source": [
        "### Step 5: Phylogenetic Tree Construction\n",
        "\n",
        "The next phase in our project involves constructing phylogenetic trees to visualize and analyze the evolutionary relationships among the species. We will use three distinct methods, each providing unique insights.\n",
        "\n",
        "#### Phylogenetic Tree Construction Methods:\n",
        "\n",
        "1. **Bayesian Inference Trees**:\n",
        "    - **Overview**: This method uses Bayesian statistics to estimate the likelihood of different evolutionary histories. It's particularly useful for its ability to estimate branch lengths and support values.\n",
        "    - **Tools**: MrBayes, BEAST\n",
        "        - MrBayes ([Official Website](https://nbisweden.github.io/MrBayes/manual.html/)) is widely recognized for its robustness in Bayesian inference.\n",
        "        - BEAST2 ([BEAST Software](https://www.beast2.org/)) is another powerful tool, offering advanced features for complex evolutionary models.\n",
        "\n",
        "2. **Maximum Likelihood Trees**:\n",
        "    - **Overview**: Maximum Likelihood methods evaluate tree topologies based on the likelihood of observed data given a tree model. It's known for its statistical rigor and accuracy.\n",
        "    - **Tools**: RAxML, PhyML\n",
        "        - RAxML ([RAxML GitHub](https://github.com/stamatak/standard-RAxML)) is preferred for large datasets due to its efficiency.\n",
        "        - PhyML ([PhyML Documentation](http://www.atgc-montpellier.fr/phyml/)) offers a balance of speed and accuracy, with a user-friendly interface.\n",
        "\n",
        "3. **Neighbor-Joining Trees**:\n",
        "    - **Overview**: The Neighbor-Joining method is a distance-based approach that constructs phylogenetic trees by evaluating the genetic distance between sequences. It is known for its speed and simplicity, making it well-suited for initial exploratory analyses.\n",
        "    - **Tools**:\n",
        "        - MEGA: A versatile tool specifically used here for constructing Neighbor-Joining trees. It's recognized for its ease of use and effectiveness in phylogenetic analysis. [MEGA Software](https://www.megasoftware.net/)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4iwUHLNzj0l7"
      },
      "outputs": [],
      "source": [
        "# TODO: Import necessary libraries for tree construction\n",
        "# from Bio import Phylo\n",
        "# from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
        "\n",
        "# TODO: Construct a phylogenetic tree using the Neighbor Joining method\n",
        "# def construct_tree_NJ(aligned_sequences):\n",
        "    # Implement the tree construction using Neighbor Joining\n",
        "\n",
        "# TODO: Repeat the process for Maximum Likelihood and Supertree methods\n",
        "# def construct_tree_ML(aligned_sequences):\n",
        "    # Implement the tree construction using Maximum Likelihood\n",
        "\n",
        "# def construct_tree_BI(gene_trees):\n",
        "    # Implement the construction of a tree using bayes inference\n",
        "\n",
        "# TODO: Visualize and save the constructed trees\n",
        "# Phylo.draw(tree, do_show=False)\n",
        "# Phylo.write(tree, 'tree_output.xml', 'phyloxml')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BWHFilRrZ0kF",
        "outputId": "9a22fe12-5d5e-4816-ebdd-d98b6fab9440"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: biopython in /usr/local/lib/python3.10/dist-packages (1.83)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.23.5)\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython\n",
        "# aligned_sequences_clustalo_oneline.fasta"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bDljeEhWZ3Vh",
        "outputId": "80b1ac1d-e162-438a-cbc5-fe9f1eea525a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "  ______ Eudynamys_scolopaceus_birds_NC_060520\n",
            " |\n",
            " |       ___ Cygnus_cygnus_birds_NC_027095\n",
            " |   ___|\n",
            " |  |   |______ Aquila_nipalensis_hawks_&_eagles_NC_0...\n",
            " | _|\n",
            " || | _ Anas_platyrhynchos_birds_NC_009684\n",
            " || ||\n",
            " ||  |_ Anser_cygnoides_birds_NC_023832\n",
            " ||\n",
            " ||  _____ Oncorhynchus_keta_bony_fishes_NC_017838\n",
            " || |\n",
            " || |   ___ Callicebus_lugens_primates_NC_024630\n",
            " ||,| ,|\n",
            " |||| ||___ Ovis_aries_sheep_NC_001941\n",
            " |||| ||\n",
            " |||| ||___ Pteropus_vampyrus_bats_NC_026542\n",
            " ||||_||\n",
            " |||  ||____ Myotis_lucifugus_bats_NC_029849\n",
            " |||  |\n",
            " |||  |  _ Mus_musculus_rodents_NC_005089\n",
            " | |  |_|\n",
            " | |    |_ Rattus_norvegicus_Rodents_NC_001665\n",
            " | |\n",
            " | |   _________ Pelecanus_crispus_birds_OR620163\n",
            " | |  |\n",
            " | |  |             _________ Damon_diadema_whip_spiders_NC_011293\n",
            " | |__|            |\n",
            " |    |            |   _____ Malaza_empyreus_moths_NC_048454\n",
            " |    |            |  |\n",
            " |    |            | ,|___ Ruspolia_dubia__crickets_NC_009876\n",
            " |    |____________| ||\n",
            " |                 | ||  ___ Corydalus_cornutus_insects_NC_011276\n",
            " |                 | || |\n",
            " |                 | ||,| __ Polystoechotes_punctatus_insects_NC_0...\n",
            " |                 | |||,|\n",
            " |                 |_|||||__ Ascaloptynx_appendiculata_insects_NC_...\n",
            " |                   ||||\n",
            " |                   | ||_________ Drosophila_melanogaster_flies_NC_024511\n",
            " |                   | |\n",
            " |                   | |____ Tribolium_castaneum_Insects_NC_003081\n",
            " |                   |\n",
            " |                   |  _________ Diadegma_semiclausum_wasps,_ants_&_be...\n",
            " |                   |_|\n",
            "_|                     |______ Apis_mellifera_Bees_NC_051932\n",
            " |\n",
            " | __ Struthio_camelus_birds_NC_002785\n",
            " ||\n",
            " ,|__ Dromaius_novaehollandiae_birds_NC_002784\n",
            " ||\n",
            " ||__ Apteryx_australis_birds_MN356385\n",
            " |\n",
            " |___ Gallus_gallus_birds_NC_053523\n",
            " |\n",
            " | _ Aptenodytes_forsteri_birds_NC_027938\n",
            " ,|\n",
            " ||_ Spheniscus_demersus_birds_NC_022817\n",
            " |\n",
            " |____________ Tyto_alba_birds_MZ318036\n",
            " |\n",
            " | __ Taeniopygia_guttata_birds_NC_007897\n",
            " ||\n",
            "  | , Corvus_frugilegus_birds_NC_002069\n",
            "  |_|\n",
            "    | Corvus_brachyrhynchos_birds_NC_026461\n",
            "\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "1"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "from Bio import AlignIO, Phylo\n",
        "from Bio.Phylo.TreeConstruction import DistanceMatrix\n",
        "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor, ParsimonyTreeConstructor\n",
        "from Bio.Phylo.TreeConstruction import DistanceCalculator, ParsimonyScorer\n",
        "# Step 1: Read sequences from the FASTA file\n",
        "fasta_file = \"/content/dataset/aligned_sequences_clustalo_oneline.fasta\"\n",
        "alignment = AlignIO.read(fasta_file, \"fasta\")\n",
        "\n",
        "calculator = DistanceCalculator(\"identity\")\n",
        "distance_matrix = calculator.get_distance(alignment)\n",
        "constructor = DistanceTreeConstructor(calculator)\n",
        "nj_tree = constructor.nj(distance_matrix)\n",
        "Phylo.draw_ascii(nj_tree)\n",
        "Phylo.write(nj_tree, 'tree_output.xml', 'phyloxml')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "h_WBklvklAqA"
      },
      "source": [
        "we use MEGA to import aligned fasta file, (we can import raw fasta files and align them using MEGA but here we import aligned sequences), then we choose to construct Makimum likelihood tree, options were set to default"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2UGu3OQjPrbA"
      },
      "source": [
        "<img src='https://drive.google.com/thumbnail?id=1vQOM-zXWq7RjlmE4dEtNw-CDAQ4xNRpE'>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eVz2641BIOE8"
      },
      "source": [
        "same steps were done to construct neighbor joining tree, except here we set a test phylogeny step to use 500 replications of bootstrap method, which according to previous tests on this subject, leads to more accurate tree\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HlTxzxitQWnJ"
      },
      "source": [
        "<img src='https://drive.google.com/thumbnail?id=1c5HsZrfs0CQ0XAbEOMi3LauQfFmuPpuM'>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NrnZMeATRi_h"
      },
      "source": [
        "for constructing bayesian inference tree, first we use aligned sequences to get nexus file, there are so many tools to do this, but here we used AliView software to get the .nex datatype, then we use this data type to infere Bayesian inference phylogenetic tree."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gvf1ZFwNRnZe"
      },
      "source": [
        "<img src='https://drive.google.com/thumbnail?id=1cgTMxY47U_nqF4JJQTvqoo1zoC-T0AnY'>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LhBLPbhyTMkx"
      },
      "source": [
        "the tool we're gonna use here is MrBayes (3.2.7 version), proceeding steps were performed in order to construct BI tree.\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "1. MrBayes > exe aligned_sequences_clustalo_oneline.nex</br>\n",
        "> Executing file \"aligned_sequences_clustalo_oneline.nex\"</br>\n",
        "  DOS line termination</br>\n",
        "  Longest line length = 46397</br>\n",
        "  Parsing file</br>\n",
        "  Expecting NEXUS formatted file</br>\n",
        "  Reading data block</br>\n",
        "  Allocated taxon set</br>\n",
        "  Allocated matrix</br>\n",
        "  Defining new matrix with 33 taxa and 46343 characters</br>\n",
        "  Data is Dna</br>\n",
        "  Gaps coded as -</br>\n",
        "  Missing data coded as ?</br>\n",
        "  Taxon  1 -> Eudynamys_scolopaceus_birds_NC_060520</br>\n",
        "  Taxon  2 -> Apis_mellifera_Bees_NC_051932</br>\n",
        "  Taxon  3 -> Pelecanus_crispus_birds_OR620163</br>\n",
        "  Taxon  4 -> Rattus_norvegicus_Rodents_NC_001665</br>\n",
        "  Taxon  5 -> Gallus_gallus_birds_NC_053523</br>\n",
        "....\n",
        "\n",
        "\n",
        "---\n",
        "2. MrBayes > lset nst=6 rates=invgamma</br>\n",
        ">  Setting Nst to 6</br>\n",
        "   Setting Rates to Invgamma</br>\n",
        "   Successfully set likelihood model parameters\n",
        "\n",
        "---\n",
        "\n",
        "3. mcmc ngen=20000 samplefreq=100 printfreq=100 diagnfreq=1000</br>\n",
        ">  Setting number of generations to 20000</br>\n",
        "   Setting sample frequency to 100</br>\n",
        "   Setting print frequency to 100</br>\n",
        "   Setting diagnosing frequency to 1000</br>\n",
        "   Running Markov chain</br>\n",
        "   MCMC stamp = 5988690643</br>\n",
        "   Seed = 762457471</br>\n",
        "   Swapseed = 1707117446</br>\n",
        "....</br>\n",
        "   19600 -- (-369057.537) (-369132.535) [-369032.121] (-369076.958) * [-369004.518] (-369008.116) (-369010.075) (-369016.818) -- 0:00:42 </br>\n",
        "   19700 -- (-369055.871) (-369134.445) [-369030.308] (-369078.326) * (-369002.091) [-369007.878] (-369008.541) (-369021.728) -- 0:00:31</br>\n",
        "   19800 -- (-369052.821) (-369128.822) [-369033.383] (-369076.015) * [-368999.426] (-369001.347) (-369009.764) (-369018.846) -- 0:00:21</br>\n",
        "   19900 -- (-369051.006) (-369125.358) [-369038.452] (-369076.178) * [-368996.576] (-369004.653) (-368999.092) (-369019.324) -- 0:00:10</br>\n",
        "   20000 -- (-369049.700) (-369125.901) [-369032.708] (-369074.678) * [-368993.184] (-368997.261) (-369004.145) (-369020.128) -- 0:00:00</br></br>\n",
        "\n",
        "   Average standard deviation of split frequencies: 0.166069</br></br>\n",
        "\n",
        "   Continue with analysis? (yes/no):</br>\n",
        "  Additional number of generations: no</br>\n",
        "\n",
        "\n",
        "---\n",
        "4. MrBayes > sump\n",
        ">   <p>Summarizing parameters in files aligned_sequences_clustalo_oneline.nex.run1.p and aligned_sequences_clustalo_oneline.nex.run2.p\n",
        "   Writing summary statistics to file aligned_sequences_clustalo_oneline.nex.pstat\n",
        "   Using relative burnin ('relburnin=yes'), discarding the first 25 % of samples. </br> You can use these\n",
        "   graphs to determine what the burn in for your analysis should be.\n",
        "   When the log probability starts to plateau you may be at station-\n",
        "   arity. Sample trees and parameters after the log probability\n",
        "   plateaus.Also examine the convergence diagnostics provided by\n",
        "   the 'sump' and 'sumt' commands for all the parameters in your\n",
        "   model. Remember that the burn in is the number of samples to dis-\n",
        "   card.</p>\n",
        "...\n",
        "\n",
        "---\n",
        "5. MrBayes > sumt\n",
        ">   Summarizing trees in files \"aligned_sequences_clustalo_oneline.nex.run1.t\" and \"aligned_sequences_clustalo_oneline.nex.run2.t\". Writing statistics to files aligned_sequences_clustalo_oneline.nex.</br>\n",
        "\n",
        "   Summary statistics for partitions with frequency >= 0.10 in at least one run:</br>\n",
        "       Average standard deviation of split frequencies = 0.166069</br>\n",
        "       Maximum standard deviation of split frequencies = 0.646230</br>\n",
        "       Average PSRF for parameter values (excluding NA and >10.0) = 1.076</br>\n",
        "       Maximum PSRF for parameter values = 1.725</br>\n",
        "...\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "B94kKvgLR4J1"
      },
      "source": [
        "<img src='https://drive.google.com/thumbnail?id=1ur-6eCDgCMjB8Gq--RIT5YUr95oydOHF'>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zydX2Rnyj0l8"
      },
      "source": [
        "### Step 6: In-Depth Phylogenetic Tree Visualization\n",
        "\n",
        "Having constructed phylogenetic trees using different methods, our next task is to visualize these trees effectively. This step is crucial for interpreting the results and communicating our findings.\n",
        "\n",
        "#### Visualization Tools:\n",
        "\n",
        "1. **FigTree**:\n",
        "    - **Overview**: FigTree is designed for the graphical representation of phylogenetic trees. It's excellent for creating publication-ready visualizations.\n",
        "    - **Resource**: [FigTree Tool](http://tree.bio.ed.ac.uk/software/figtree/)\n",
        "    - **Usage**: Use FigTree to add detailed annotations, adjust branch colors, and format tree layouts for clear, interpretable visualizations.\n",
        "\n",
        "2. **iTOL (Interactive Tree Of Life)**:\n",
        "    - **Overview**: iTOL is a web-based tool for the display, annotation, and management of phylogenetic trees, offering extensive customization options.\n",
        "    - **Resource**: [iTOL Website](https://itol.embl.de/)\n",
        "    - **Usage**: Ideal for interactive tree visualizations. It allows users to explore different layers of data through their tree, such as adding charts or color-coding branches.\n",
        "\n",
        "3. **Dendroscope**:\n",
        "    - **Overview**: Dendroscope is a software program for viewing and editing phylogenetic trees, particularly useful for large datasets.\n",
        "    - **Resource**: [Dendroscope Download](https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/dendroscope/)\n",
        "    - **Usage**: Utilize Dendroscope when dealing with large and complex trees or when you need to compare multiple trees side-by-side.\n",
        "\n",
        "#### Task:\n",
        "\n",
        "- **Visualize Each Tree**: Use one or more of the above tools to visualize the phylogenetic trees you constructed using Bayesian inference, maximum likelihood, and neighbor-joining methods.\n",
        "- **Highlight Differences**: Focus on highlighting the differences and similarities between the trees obtained from the different methods. Pay attention to tree topology, branch lengths, and any notable patterns.\n",
        "- **Interpretation and Presentation**: Aim for visualizations that are not only accurate but also interpretable and visually appealing. This will enhance the clarity of your work."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7twd3HBxSbXT"
      },
      "source": [
        "visualizing maximum likelihood tree using figtree:"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yTLMOPs2Siki"
      },
      "source": [
        "<img src='https://drive.google.com/thumbnail?id=1OBpyIly9GtAEVaKxwnmRYjJ-utzvz-xI'>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GBEL4544TRdf"
      },
      "source": [
        "visualizing neighbor joining tree using figtree:"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "brO07i-tTks9"
      },
      "source": [
        "<img src='https://drive.google.com/thumbnail?id=1REJVSDY0JGn1zId73CzOX9gjQnpgkj8w'>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "naX6_CwdUE_H"
      },
      "source": [
        "visualizing bayesian inference tree using figtree:"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BB7dx2QWUJ87"
      },
      "source": [
        "<img src='https://drive.google.com/thumbnail?id=14NtmOAhDhdl8IjkzD-2tTecOMzwced5M'>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aogiisVXj0l8"
      },
      "source": [
        "### Cross-Disciplinary Applications (Optional)\n",
        "\n",
        "This is an optional part with bonus, relative to the depth of your analysis. Refer to the first part of this notebook. You have complete freedom to do this part anyway you like, but to gain a portion of the bonus score for this section, a bare minimum effort is required."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xAIUqqioj0l8"
      },
      "source": [
        "### Conclusion and Reflective Insights\n",
        "\n",
        "As we conclude our exploration of phylogenetic tree construction and analysis, let's reflect on the insights learned from this task and consider questions that emerge from our findings.\n",
        "\n",
        "#### Interpretation of Results:\n",
        "\n",
        "- Reflect on the phylogenetic trees produced by each method (Bayesian inference, maximum likelihood, and neighbor-joining). Consider how the differences in tree topology might offer varied perspectives on the evolutionary relationships among the species.\n",
        "\n",
        "#### Questions to Ponder:\n",
        "\n",
        "1. **Species Divergence**: Based on the trees, which species appear to have the most ancient divergence? How might this information contribute to our understanding of their evolutionary history?\n",
        "   \n",
        "2. **Common Ancestors**: Are there any unexpected pairings or groupings of species that suggest a closer evolutionary relationship than previously thought? How could this reshape our understanding of these species' evolutionary paths?\n",
        "\n",
        "3. **Methodology Insights**: Considering the discrepancies between the trees generated by different methods, what might this tell us about the limitations and strengths of each phylogenetic analysis method?\n",
        "\n",
        "4. **Conservation Implications**: Considering the evolutionary relationships revealed in your phylogenetic analysis, what insights can be gained for conservation strategies? Specifically, how could understanding the close evolutionary ties between species, which might be facing distinct environmental challenges, guide targeted conservation efforts?\n",
        "\n",
        "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
        "    <strong>Don't forget to answer these questions!</strong>\n",
        "</blockquote>\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KjrVFpkdE-lx"
      },
      "source": [
        "**Species Divergence**\n",
        "considering the phylogenetic trees and the times corresponding each specie, \"Pelecanus crispus\" is the most ancienct specie among the species in our dataset.\n",
        "Understanding the most ancient divergences among species provides useful insights into evolutionary history. By identifying basal positions on phylogenetic trees, scientists can root the tree of life, construct evolutionary timelines, and unravel biogeographic events.</br>\n",
        "**Common Ancestors**\n",
        "the only grouping that's a bit different is the grouping of species sheep, primate, bat and rodent with bony fishes, in a way that it seems like the first group which are all mammalian have same common ancestor with a fish, which is unexpected. my personal thought is that we infer from this result that there have been interspecies connections which had led to creation of new species.\n",
        "It suggests that key innovations, such as terrestrial locomotion in mammals, may have deeper roots in vertebrate evolution. Understanding this shared ancestry provides insights into both constraints and opportunities that have shaped their evolutionary trajectories.</br>\n",
        "**Methodology Insights**\n",
        "Neighbor joining is clustering algorithms that can make quick trees but is not the most reliable, especially when dealing with deeper divergence times. This method is good to give you an overall idea about your data, but is almost never acceptable for publication. Neighbor joining is most vulnerable to missing data, Maximum likelihood is the least vulnerable inference. The reason is simple: the more missing data, the less the distance matrix is a reflection of the true tree. <br/>\n",
        "Maximum likelihood and Bayesian methods can apply a model of sequence evolution and are ideal for building a phylogeny using sequence data. These methods are the two methods that are most often used in publications. The main downside of these methods is that they are computational expensive.\n",
        "If the phylogeny is the main focus of your work, my suggestion is to make both maximum likelihood and Bayesian trees. For these methods you will need to choose a model of sequence evolution.<br/>\n",
        "The lower the diversity in your data, the less discriminative Maximum likelihood and Bayesian will be, you won't get necessarily wrong trees, but random ones. Maximum likelihood will still choose one tree, which is however not significantly better than other alternatives and may include numerous branches that receive lower BS support than conflicting alternatives.<br/>\n",
        "**Conservation Implications**\n",
        "Understanding the close evolutionary relationships between species helps us develop conservation strategies that benefit multiple related species. By identifying key species that represent evolutionary diversity, we can prioritize their protection to safeguard the broader evolutionary heritage. This approach, known as cross-species conservation, allows us to leverage shared ecological requirements and genetic traits. Additionally, recognizing these evolutionary ties informs adaptive management strategies, such as translocations, which enhance genetic diversity and resilience to environmental changes. Ultimately, integrating evolutionary perspectives into conservation planning improves the efficiency and effectiveness of conservation actions, ensuring the preservation of biodiversity and ecosystem integrity in the face of environmental threats."
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.7"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
